File: dataset_reduzido/avatar/Python/Code/atcoder_ABC155_D.py
N , K = map ( int , input ( ).split ( ) )
A = list ( map ( int , input ( ).split ( ) ) )
A.sort ( )
ll = - 10 ** 18 - 1
rr = 10 ** 18 + 1
while ll + 1 < rr :
    x = ( ll + rr ) // 2
    tot = 0
    for i in range ( N ) :
        if A [ i ] < 0 :
            l = - 1
            r = N
            while l + 1 < r :
                c = ( l + r ) // 2
                if A [ i ] * A [ c ] < x :
                    r = c
                else :
                    l = c
            tot += N - r
        else :
            l = - 1
            r = N
            while l + 1 < r :
                c = ( l + r ) // 2
                if A [ i ] * A [ c ] < x :
                    l = c
                else :
                    r = c
            tot += r
        if A [ i ] * A [ i ] < x :
            tot -= 1
    tot //= 2
    if tot < K :
        ll = x
    else :
        rr = x
print ( ll )

File: output/LLaMa_repaired_1/avatar/Python/Java/atcoder_ABC155_D.java
import java.util.Arrays;
import java.util.Scanner;

public class atcoder_ABC155_D {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int K = scanner.nextInt();
        int[] A = new int[N];
        for (int i = 0; i < N; i++) {
            A[i] = scanner.nextInt();
        }
        Arrays.sort(A);
        long ll = -1L * (long) (1e18);
        long rr = 1L * (long) (1e18 + 1);
        while (ll + 1 < rr) {
            long x = (ll + rr) / 2;
            long tot = 0;
            for (int i = 0; i < N; i++) {
                if (A[i] < 0) {
                    int l = -1;
                    int r = N;
                    while (l + 1 < r) {
                        int c = (l + r) / 2;
                        if (A[i] * A[c] < x) {
                            r = c;
                        } else {
                            l = c;
                        }
                    }
                    tot += Math.max(N - r, 0);
                    tot--;
                } else {
                    int l = -1;
                    int r = N;
                    while (l + 1 < r) {
                        int c = (l + r) / 2;
                        if (A[i] * A[c] < x) {
                            l = c;
                        } else {
                            r = c;
                        }
                    }
                    tot += r;
                }
                if (A[i] != 0 && A[i] * A[i] < x) {
                    tot--;
                }
            }
            tot /= 2;
            if (tot < K) {
                ll = x;
            } else {
                rr = x;
            }
        }
        System.out.println(ll);
    }
}